hcl2template/parser.go

```diff
@@ -293,111 +272,314 @@ func filterVarsFromLogs(inputOrLocal Variables) {
+	attrs, moreDiags := block.Body.JustAttributes()
 	diags = append(diags, moreDiags...)
+	for key, attr := range attrs {
+		moreDiags = cfg.InputVariables.decodeVariable(key, attr, ectx)
+		diags = append(diags, moreDiags...)
+	}
 
-	for _, block := range content.Blocks {
-		switch block.Type {
-		case dataSourceLabel:
-			datasource, moreDiags := p.decodeDataBlock(block)
-			diags = append(diags, moreDiags...)
-			if moreDiags.HasErrors() {
-				continue
-			}
-			ref := datasource.Ref()
-			if existing, found := cfg.Datasources[ref]; found {
-				diags = append(diags, &hcl.Diagnostic{
-					Severity: hcl.DiagError,
-					Summary:  "Duplicate " + dataSourceLabel + " block",
-					Detail: fmt.Sprintf("This "+dataSourceLabel+" block has the "+
-						"same data type and name as a previous block declared "+
-						"at %s. Each "+dataSourceLabel+" must have a unique name per builder type.",
-						existing.block.DefRange.Ptr()),
-					Subject: datasource.block.DefRange.Ptr(),
-				})
-				continue
-			}
-			if cfg.Datasources == nil {
-				cfg.Datasources = Datasources{}
-			}
-			cfg.Datasources[ref] = *datasource
-		}
+	return diags
+}
+
+// decodeBuildBlock shallowly decodes a build block from the config.
+//
+// The final decoding step (which requires an up-to-date context) will be done
+// when we need it.
+func (cfg *PackerConfig) decodeBuildBlock(block *hcl.Block) hcl.Diagnostics {
+	build := &BuildBlock{
+		block: block,
+	}
+
+	cfg.Builds = append(cfg.Builds, build)
+
+	return nil
+}
+
+func (cfg *PackerConfig) decodeSourceBlock(block *hcl.Block) hcl.Diagnostics {
+	source, diags := cfg.decodeSource(block)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	if cfg.Sources == nil {
+		cfg.Sources = map[SourceRef]SourceBlock{}
 	}
 
+	ref := source.Ref()
+	if existing, found := cfg.Sources[ref]; found {
+		return append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Duplicate " + sourceLabel + " block",
+			Detail: fmt.Sprintf("This "+sourceLabel+" block has the "+
+				"same builder type and name as a previous block declared "+
+				"at %s. Each "+sourceLabel+" must have a unique name per builder type.",
+				existing.block.DefRange.Ptr()),
+			Subject: source.block.DefRange.Ptr(),
+		})
+	}
+
+	cfg.Sources[ref] = source
+
 	return diags
 }
+
+// decodeBuildSource reads a used source block from a build:
+//
+//	build {
+//	  source "type.example" {
+//	    name = "local_name"
+//	  }
+//	}
+func (cfg *PackerConfig) decodeBuildSource(block *hcl.Block) (SourceUseBlock, hcl.Diagnostics) {
+	ref := sourceRefFromString(block.Labels[0])
+	out := SourceUseBlock{SourceRef: ref}
+	var b struct {
+		Name string   `hcl:"name,optional"`
+		Rest hcl.Body `hcl:",remain"`
+	}
+	diags := gohcl.DecodeBody(block.Body, nil, &b)
+	if diags.HasErrors() {
+		return out, diags
+	}
+	out.LocalName = b.Name
+	out.Body = b.Rest
+	return out, nil
+}
+
+func (source *SourceBlock) finalizeDecodeSource(cfg *PackerConfig) hcl.Diagnostics {
```

Comment:
Since this method is on the SourceBlock we can drop the Source suffix. 
```suggestion
func (source *SourceBlock) finalizeDecode(cfg *PackerConfig) hcl.Diagnostics {
```